### DISCUSSION TITLE
multiple compilers on BindParam

### DISCUSSION THREAD

**Discussion by DeoLeung** [sqlalchemy/sqlalchemy#10286] [Usage Questions] [answered]
hi, using `1.4` , 

we have different teams building libraries, in one library

```python
@compiles(BindParameter)
def _render_literal_bindparam(element, compiler, use_my_literal_recipe=False,
                              **kw):
  ...
```

in another library/app there is 
```python
@compiles(BindParameter, 'asyncpg')
def _render_literal_bindparam1(element,
                              compiler,
                              compile_bindparam_as_dollar=False,
                              **kw):
  ...
```

it seems depends import order, either will be overwritten, is there a way to keep both, or some way to warn it? it's kind of hard to debug :)

**Comment by CaselIT** [comment]
Hi,

If you import both I think the invocation rule should be stable.
Do you have issues even if you import both?

**Comment by zzzeek** [comment, marked as answer]
well first weird thing is that `@compiles` rules dont key off of the driver name, in this case `asyncpg`, they only take effect for the driver name, that is `postgresql` (I just checked the source to make sure).  So right off if someone has `asyncpg` like that, I think that compiler hook does nothing.

so maybe your other team's recipe is not doing anything anyway and you can just remove it.

Next thing, if you had one rule against nothing, and the other against "postgresql", that's fine also, the one that's against nothing is the default and would not take place when postgresql is in use.

as far as having both methods cascade onto one another, we don't currently have explicit support for this feature yet (but you can of course organize this on your end using a third hook that mediates the other two).    

here is the bigger picture (besides the non working "asyncpg" dialect target) - overriding `@compiles` for `BindParameter` is a pretty squirrelly use case, having to do it once is way out there, doing it twice in one application seems worth revisiting what use case is actually happening here.

**Comment by DeoLeung** [comment]
yes, the name 'asyncpg' does not work :(

currently we do something like following to ensure order

```python
@compiles(BindParameter)
def _render_literal_bindparam(element, compiler, use_my_literal_recipe=False,
                              **kw):
  ...
in another library/app there is

from a import _render_literal_bindparam

@compiles(BindParameter, 'postgresql')
def bindparam_asyncpg(element,
                              compiler,
                              compile_bindparam_as_dollar=False,
                              **kw):
  ...
  return _render_literal_bindparam(element, compiler, **kw)
```

**Reply by CaselIT** [reply]
Why do you need to do that? What's the issue with the default way sqlalchemy uses bind parameters?

**Reply by DeoLeung** [reply]
we were using asyncpg directly (not sqlalchemy engine) to execute raw sql, which compiles sql from sqlalchemy statement.

like `.where(table.c.a == bindparam('$1')`
we want something like `where a = $1`

**Reply by CaselIT** [reply]
If you upgrade to 2.0 thats supported directly
